lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java

```diff
@@ -389,4 +501,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
         }
         return Domain.create(ValueSet.all(type), hasNullValue);
     }
+
+    // rangeStatistics is neither null nor empty, checked by the caller
+    private static <F, T extends Comparable<T>> Domain createDomain(Type type,
+                                                                    boolean hasNullValue,
+                                                                    List<ParquetRangeStatistics<F>> rangeStatistics,
+                                                                    Function<F, T> function)
+    {
+        Range firstRange = null;
+        Range[] restRanges = new Range[rangeStatistics.size() - 1];
+        for (int i = 0; i < rangeStatistics.size(); i++) {
+            F min = rangeStatistics.get(i).getMin();
+            F max = rangeStatistics.get(i).getMax();
+            Range range;
+            if (min != null && max != null) {
+                range = Range.range(type, function.apply(min), true, function.apply(max), true);
+            }
+            else if (max != null) {
+                range = Range.lessThanOrEqual(type, function.apply(max));
+            }
+            else if (min != null) {
+                range = Range.greaterThanOrEqual(type, function.apply(min));
+            }
+            else {
+                return Domain.create(ValueSet.all(type), hasNullValue);
```

Comment:
This could use an explanatory comment. Why does it return early only in this case?